compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1479,9 +1479,27 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
     /** Like tp1 <:< tp2, but returns false immediately if we know that
      *  the case was covered previously during subtyping.
+     *
+     *  A type has been covered previously in subtype checking if it
+     *  is some combination of TypeRefs that point to classes, where the
+     *  combiners are AppliedTypes, RefinedTypes, RecTypes, And/Or-Types or AnnotatedTypes.
+     *
+     *  The exception is that if both sides contain OrTypes, the check hasn't been covered.
+     *  See #17465.
      */
     def isNewSubType(tp1: Type): Boolean =
-      if (isCovered(tp1) && isCovered(tp2))
+      def isCovered(tp: Type): CoveredStatus =
+        tp.dealiasKeepRefiningAnnots.stripTypeVar match
+          case tp: TypeRef if tp.symbol.isClass && tp.symbol != NothingClass && tp.symbol != NullClass => CoveredStatus.Covered
```

Comment:
```suggestion
          case tp: TypeRef =>
            if tp.symbol.isClass && tp.symbol != NothingClass && tp.symbol != NullClass
            then CoveredStatus.Covered
            else CoveredStatus.Uncovered
```
It's a bit faster since it does not try the other patterns in case of a non-class TypeRef.